experimental/lib/Transforms/LSQSizing/HandshakeSizeLSQs.cpp

```diff
@@ -0,0 +1,661 @@
+  // candidate and that the start time of the candidate is not zero
+  insertStartnodeShiftingEdges(graph, startNode, startTimes);
+
+  // Calculate the earliest start times for each operation in the graph. This is
+  // needed to find the allocation times
+  graph.setEarliestStartTimes(startNode);
+
+  std::vector<unsigned> IIs;
+
+  // The collisions argument can be passed in the compile script
+  // For now there are 3 defined cases: full, half, none
+  // full: determines the worst case II, for the case that all lsq ops have
+  // memory collisions
+  // half: determines the worst case II and alternate it with
+  // the best case II (from buffer placement), for the case that half of the lsq
+  // ops have memory collisions
+  // none: gets the best case II(from buffer
+  // placement), for the case that no lsq ops have memory collisions
+  //
+  // It would also easily be possible to add more cases, by just pushing the IIs
+  // into the IIs vector in the order they are occuring
+
+  if (collisions == "full") {
+    IIs.push_back(graph.getWorstCaseII());
+  } else if (collisions == "half") {
+    IIs.push_back(initialII);
+    IIs.push_back(graph.getWorstCaseII());
+  } else {
+    // none is the default case
+    IIs.push_back(initialII);
+  }
+
+  std::unordered_map<unsigned, TimePerOpMap> loadAllocTimes;
+  std::unordered_map<unsigned, TimePerOpMap> storeAllocTimes;
+  std::unordered_map<unsigned, TimePerOpMap> loadDeallocTimes;
+  std::unordered_map<unsigned, TimePerOpMap> storeDeallocTimes;
+
+  for (auto &II : IIs) {
+    graph.setNewII(II);
+    loadAllocTimes.insert_or_assign(
+        II, getAllocTimes(graph, startNode, loadOps, phiNodes));
+    storeAllocTimes.insert_or_assign(
+        II, getAllocTimes(graph, startNode, storeOps, phiNodes));
+    loadDeallocTimes.insert_or_assign(
+        II, getLoadDeallocTimes(graph, startNode, loadOps));
+    storeDeallocTimes.insert_or_assign(
+        II, getStoreDeallocTimes(graph, startNode, storeOps));
+  }
+  // Get Load and Store Sizes
+  SizePerOpMap loadSizes = calcQueueSize(loadAllocTimes, loadDeallocTimes, IIs);
+  SizePerOpMap storeSizes =
+      calcQueueSize(storeAllocTimes, storeDeallocTimes, IIs);
+
+  LSQSizingResult result;
+  for (auto &entry : loadSizes) {
+    unsigned storeSize = storeSizes.find(entry.first) != storeSizes.end()
+                             ? storeSizes[entry.first]
+                             : 0;
+    result.insert({entry.first, std::make_tuple(entry.second, storeSize)});
+  }
+
+  for (auto &entry : storeSizes) {
+    if (result.find(entry.first) == result.end())
+      result.insert({entry.first, std::make_tuple(0, entry.second)});
+  }
+
+  return result;
+}
+
+std::tuple<mlir::Operation *, StartTimes>
+HandshakeSizeLSQsPass::findStartTimes(AdjListGraph graph) {
+  StartTimes startTimes;
+
+  // Find all mux and control_merge ops as candidates for start node
+  std::vector<mlir::Operation *> muxOps =
```

Comment:
Yep, even more "LLVM-style".

```cpp
std::vector<mlir::Operation *> startNodeCandidates;
llvm::copy(funcOp.getOps<handshake::MergeLikeOpInterface>(), std::back_inserter(startNodeCandidates);
```

Reply:
Here again, it is a problem that I only want the Operations of these types inside the specific CFDFC and not of the whole FuncOp. But I created a templated function for my graph, so now it would look like this:
 `  std::vector<mlir::Operation *> startNodeCandidates =
      graph.getOperationsWithOpType<handshake::MergeLikeOpInterface>();`
I think this should also be fine.